exercises/largest-series-product/examples/success-standard/src/Series.hs

```diff
@@ -1,22 +1,34 @@
-module Series (Error(..), largestProduct) where
+module Series
+  ( Error(..)
+  , largestProduct
+  )
+where
 
-import Control.Monad    ((>=>))
-import Data.Char        (digitToInt, isDigit)
-import Data.List        (tails)
-import Data.Maybe       (mapMaybe)
-import Safe             (maximumMay)
-import Safe.Exact       (takeExactMay)
+import           Data.Char
+-- for window':  import           Data.List                      ( tails )
 
-data Error = InvalidSpan | InvalidDigit Char deriving (Show, Eq)
+data Error
+  = InvalidSpan
+  | InvalidDigit Char
+  deriving (Show, Eq)
 
-rightOr :: a -> Maybe b -> Either a b
-rightOr a Nothing = Left a
-rightOr _ (Just x) = Right x
+largestProduct :: Int -> String -> Either Error Integer
+largestProduct size digits = if size > length digits || size < 0
+  then Left InvalidSpan
+  else do
+    digits' <- traverse parseDigits digits
```

Comment:
Excellent idea to handle errors monadically. I might have used `mapM` instead of `traverse` now that we're in a do-block, but since they're equivalent, I don't mind. I suppose I like the notion of `mapM` because of the `...M` suffix.

If, however, you put this in a separate function, e.g. `digits :: String -> Either Error [Int]`, you can compose it more easily.